### ISSUE TITLE
Pros and cons of a generic packetizer vs. codec specific packetizer should be documented

### ISSUE THREAD

**Issue by youennf** [murillo128/codec-agnostic-rtp-payload-format#4]
As discussed at last IETF meeting, there were concerns about a generic packetizer, for instance in terms of error resilience. It would be good to document this in the draft.
My understanding is that with correct frame splitting (per spatial/temporal/quality layer) and sufficient header extension information, the generic packetizer could be perf neutral while being more convenient to use.

**Comment by murillo128** [comment]
There are some codec specific features that would be lost,  like slicing/tiling, DON or interleaving, although I have never seen them implemented in RTC scenarios and I see it difficult to get them working with SFUs (which is our main use case)

**Comment by youennf** [comment]
Not sure what DON or interleaving are.

AIUI, slicing could be implemented with a generic packetizer similarly to spatial layers: they would need to be processed as separate 'frames' (separately encrypted for instance) and some RTP header extensions would need to provide information about them.

The more we split a video frame in packetised frames, the more we might end up with RTP packets of small size. The generic packetization might introduce some overhead.

**Comment by murillo128** [comment]
DON is the decode order number, when the transmit order is not the decoder order (b-frames for example).

Interleaving happens typically with slicing/tiling when you can encode each slice/tile asynchronously on different threads/encoders, so instead of sending the parts of the slices in order, you interleave the results of each slice so you don't have to synchronize between them.

My main concern about slicing while it may be useful in some scenarios, SFUs will not handle them at all, so the theoretical benefit is reduced a lot. Also, it is currently not implemented in any webrtc implementation at all, so why add the complezity to support it if no one is ever going to use it?

**Comment by youennf** [comment]
> Also, it is currently not implemented in any webrtc implementation at all, so why add the complezity to support it if no one is ever going to use it?
Agreed it is out of scope of this work to try solving those issues.
It would still be good to validate that there would be a path forward though on top of the generic packetization approach.
That seems to be the case to me as long as the application doing the split handles it and provides the necessary RTP header extensions that would be needed.

I think it is worth adding some lines to the draft about those issues.
This could say something like:
- Splitting of a full frame is done between encoder and packetizer. It is the job of the application to do it, in particular prior encryption. That may allow the decoder to partially decode the content. The application should provide the necessary information outside of the content (as content may be encrypted), typically through RTP header extensions.
- Splitting full frames, and not relying on inspecting content, may introduce an overhead compared to a codec-specific packetization as more information needs to be added in RTP header extensions. This overhead should be identified in reasonable use cases. In general, even though this overhead might not be big, applications should choose generic packetization over codec-specific packetization carefully.

During the IETF meeting, issues around error resilience were also mentioned IIRC, but I do not really see the exact problems there, as long as the granularity is changed to the "splitted units" instead of NAL units and there is sufficient information about the splitted units nature in RTP header extensions.

**Comment by murillo128** [comment]
the problem is that we should need to define what are those "split" units in each codec so both the sender and receiver know how to apply the "partial decoding".  

In principle I am against adding functionalities to the spec that no one has ever implemented, so I would prefer that the people actually pushing to have this included shows some evidence about running code and what are the potential gains.

**Comment by youennf** [comment]
> we should need to define what are those "split" units

We will need to define that anyway for the cases we do care.
I think that for SFrame and SVC, the idea would be that we would split spatial layers for instance.
For non SVC, the unit would be the whole frame.

**Comment by murillo128** [comment]
> I think that for SFrame and SVC, the idea would be that we would split spatial layers for instance.
For non SVC, the unit would be the whole frame.

I agree to that, what I was meaning is that I would not go lower than the spatial layer.

If we do that, the DD already supports the packetization of the spatial layer frame into several rtp packets (with the start/end bits). If not doing SVC, we would not use the DD, but it we don't support slices/tiles, then the packetization would be trivial and no extra information would be required.
